pkg/volume/glusterfs/glusterfs.go

```diff
@@ -479,22 +510,127 @@ func (d *glusterfsVolumeDeleter) Delete() error {
 	return nil
 }
 
+func (r *glusterfsVolumeProvisioner) ConstructBitmap() error {
+
+	pvList, err := r.plugin.host.GetKubeClient().Core().PersistentVolumes().List(v1.ListOptions{LabelSelector: labels.Everything().String()})
+	if err != nil {
+		glog.Errorf("glusterfs: failed to get existing persistent volumes")
+		return fmt.Errorf("glusterfs: failed to get existing persistent volumes")
+	}
+	// Collect the SC names and allocated GIDs from existing PVs.
+	scNameGids := make(map[string][]string)
+	for _, pv := range pvList.Items {
+		if pv.Annotations[annDynamicallyProvisioned] == glusterfsPluginName {
+			scNameInAnnot := pv.Annotations[annStorageClass]
+			scNameGids[scNameInAnnot] = append(scNameGids[scNameInAnnot], pv.Annotations[volumehelper.VolumeGidAnnotationKey])
+		}
+	}
+	glog.V(3).Infof("glusterfs: PVs with SC name and allocated GID Map :%v ", scNameGids)
+
+	for key := range scNameGids {
+		scdetails, _ := r.plugin.host.GetKubeClient().Storage().StorageClasses().Get(key)
+		if scdetails.Parameters["gidMin"] == "" {
+			scdetails.Parameters["gidMin"] = defaultGidMin
+		}
+		if scdetails.Parameters["gidMax"] == "" {
+			scdetails.Parameters["gidMax"] = defaultGidMax
+		}
+		blockStartnew32, err := r.ConvertGid(scdetails.Parameters["gidMin"])
+		if err != nil {
+			glog.Errorf("glusterfs: failed to convert gid %s", scdetails.Parameters["gidMin"])
+		}
+		blockEndnew32, err := r.ConvertGid(scdetails.Parameters["gidMax"])
+		if err != nil {
+			glog.Errorf("glusterfs: failed to convert gid %s", scdetails.Parameters["gidMax"])
+		}
+		// For each SC, create a pool and store it in plugin allocator map.
+		if r.plugin.gidAllocators[key] == nil {
+			r.plugin.gidAllocators[key], err = gidPoolCreate(blockStartnew32, blockEndnew32)
+			if err != nil {
+				glog.Errorf("glusterfs: gid pool creation failed %v", err)
+				return fmt.Errorf("glusterfs: gid pool creation failed: %v.", err)
+			}
+		}
+	}
+
+	// For each SCs in the allocator, set bitmap for existing/previosly allocated GIDs.
+	for name, gids := range scNameGids {
+		for _, gid := range gids {
+			tempGid, err := r.ConvertGid(gid)
+			if err == nil {
+				if r.plugin.gidAllocators[name] != nil {
+					err = r.plugin.gidAllocators[name].Allocate(Block{Start: tempGid, End: tempGid})
+				}
+			} else {
+				glog.Errorf("glusterfs: failed to convert gid %s", gid)
+			}
+		}
+	}
+	return nil
+}
+
 func (r *glusterfsVolumeProvisioner) Provision() (*v1.PersistentVolume, error) {
```

Comment:
nevermind, a new glusterfsVolumeProvisioner is created per provision operation... I'm still not a fan of mutating the struct vars instead of passing parameters, but it's not unsafe, as long as we don't mutate anything in the shared `plugin` var